feat: add .agents skill for add-new-crd (#3837)
kubeflow/training-operator: .agents/skills/add-new-crd/SKILL.md

---
name: add-new-crd
description: Scaffolds a new Kubernetes Custom Resource Definition (CRD) for the Kubeflow Trainer project. Covers Go type definitions with code generation markers, scheme registration, webhook and controller scaffolding, RBAC markers, testing wrappers, integration tests, and manifest generation. Use when asked to add a new CRD, create a new API resource, or define a new Kubernetes custom type for Trainer.
---

> # Add a New CRD

Reviewer: Discoverability gap: `AGENTS.md` (which `CLAUDE.md` symlinks to) has no reference to `.agents/skills/`, and `CLAUDE.md`'s repo map doesn't list `.agents/`. Since `CLAUDE.md` names `AGENTS.md` as the source of truth for dev workflows, nothing will point agents here. Please add a short "Skills" pointer in `AGENTS.md` and a `.agents/` entry in the repo map.

Reviewer: That should be fine, we will fix it later via CI.

Step-by-step workflow for adding a new Custom Resource Definition to
Kubeflow Trainer. This covers type definitions, scheme registration,
webhooks, controller, tests, and code generation.

Copy this checklist and track your progress:

```
CRD Progress:
- [ ] Step 1: Define API types
- [ ] Step 2: Register types with the scheme
- [ ] Step 3: Run code generation
- [ ] Step 4: Add webhooks
- [ ] Step 5: Add controller
- [ ] Step 6: Add testing wrappers
- [ ] Step 7: Add integration tests
- [ ] Step 8: Generate manifests and verify
```

## Key files to read first

> 

Reviewer: Could we also copy the SIG arch docs to the `./agents/docs/api-conventions.md` and `./agents/docs/api_changes.md` and ensure this skill reads it?
https://github.com/kubernetes/community/tree/main/contributors/devel/sig-architecture

We might also want to add `/review-api` skill in the future to ensure APIs are following the conventions.

- **Existing type definitions**: [pkg/apis/trainer/v1alpha1/trainjob_types.go](../../../pkg/apis/trainer/v1alpha1/trainjob_types.go)
- **Scheme registration**: [pkg/apis/trainer/v1alpha1/groupversion_info.go](../../../pkg/apis/trainer/v1alpha1/groupversion_info.go)
- **Webhook setup**: [pkg/webhooks/setup.go](../../../pkg/webhooks/setup.go)
- **Controller setup**: [pkg/controller/setup.go](../../../pkg/controller/setup.go)
- **Package-level markers**: [pkg/apis/trainer/v1alpha1/doc.go](../../../pkg/apis/trainer/v1alpha1/doc.go)

## Step 1: Define API types

Create `pkg/apis/trainer/v1alpha1/<newresource>_types.go`.

### Required markers on the root type

For a **namespaced** resource:

```go
// +genclient
// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
// +kubebuilder:object:root=true

> // +kubebuilder:subresource:status

Reviewer: These two aren't universally required, so "Required markers" overstates it: `TrainingRuntime` omits `+kubebuilder:subresource:status` (no Status field) and `OptimizationJob` omits `+kubebuilder:storageversion`. Worth a line saying to omit `subresource:status` when there's no Status field, and that `storageversion` is a no-op while `v1alpha1` is the only version.

Also consider adding the two `printcolumn` markers TrainJob uses — new CRDs generally want `State`/`Age` columns.

// +kubebuilder:storageversion
type NewResource struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`
    Spec              NewResourceSpec   `json:"spec,omitempty"`
    Status            NewResourceStatus `json:"status,omitempty"`
}
```

For a **cluster-scoped** resource, add these markers:

```go
// +genclient
// +genclient:nonNamespaced
// +kubebuilder:resource:scope=Cluster
```

### Required markers on the List type

```go
// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object

> // +resource:path=newresources
> // +kubebuilder:object:root=true

Reviewer: `+resource:path=` is a dead `apiserver-builder` marker consumed by neither `controller-gen` nor `code-generator`. It lingers on older types but the newest, `OptimizationJobList`, correctly omits it — please don't teach it.

```suggestion
// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
// +kubebuilder:object:root=true
```

    Items           []NewResource `json:"items"`
}
```

### Field-level validation

Always use [CEL validation](https://kubernetes.io/docs/reference/using-api/cel/)
when applicable. Common markers:

```go
// Immutable field:
// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="field is immutable"

// Optional field with default:
// +optional
// +kubebuilder:default="value"

// List types for strategic merge patch:
// +listType=map
// +listMapKey=name

// Required field:
// +required
```

Reference existing types for the full set of markers used in this project:
- [trainjob_types.go](../../../pkg/apis/trainer/v1alpha1/trainjob_types.go) (namespaced, with printcolumns and CEL)
- [trainingruntime_types.go](../../../pkg/apis/trainer/v1alpha1/trainingruntime_types.go) (both namespaced and cluster-scoped)

## Step 2: Register types with the scheme

File: `pkg/apis/trainer/v1alpha1/groupversion_info.go`

Add both the root type and list type to `addKnownTypes()`:

```go
func addKnownTypes(scheme *apiruntime.Scheme) error {
    scheme.AddKnownTypes(SchemeGroupVersion,
        // ... existing types ...
        &NewResource{},
        &NewResourceList{},
    )

>     return metav1.AddToGroupVersion(scheme, SchemeGroupVersion)

Reviewer: `addKnownTypes` uses the unaliased `runtime` import and the canonical `GroupVersion`; the final call isn't returned (`groupversion_info.go:42-55`).

```suggestion
func addKnownTypes(scheme *runtime.Scheme) error {
    scheme.AddKnownTypes(GroupVersion,
        // ... existing types ...
        &NewResource{},
        &NewResourceList{},
    )
    metav1.AddToGroupVersion(scheme, GroupVersion)
    return nil
}
```

}
```

No changes needed in `cmd/trainer-controller-manager/main.go` - the
scheme is already registered globally via `trainer.AddToScheme(scheme)`.

## Step 3: Run code generation

```bash
make generate
```

This runs four steps:
1. `controller-gen object` - generates `zz_generated.deepcopy.go`
2. `hack/update-codegen.sh` - generates defaults, clients, informers, listers, apply configs under `pkg/client/`, and OpenAPI specs
3. `controller-gen object` for config API

> 4. Python API generation from swagger.json

Reviewer: `make generate` has prerequisites `go-mod-download manifests helm-docs` (`Makefile:194`), so Step 8's `make manifests` is already covered here. Worth flagging the Docker dependency too, since it fails late.

```suggestion
This runs (`Makefile:194`) — note `manifests` is a prerequisite, so Step 8 is already covered:
1. prerequisites `go-mod-download`, `manifests`, `helm-docs`
2. `controller-gen object` - generates `zz_generated.deepcopy.go`
3. `hack/update-codegen.sh` - generates defaults, clients, informers, listers, apply configs under `pkg/client/`, and OpenAPI specs
4. `controller-gen object` for config API
5. `hack/python-api/gen-api.sh` - Python API from swagger.json (**requires Docker/Podman**)
```

```go
// +kubebuilder:webhook:path=/mutate-trainer-kubeflow-org-v1alpha1-newresource,mutating=true,failurePolicy=fail,sideEffects=None,groups=trainer.kubeflow.org,resources=newresources,verbs=create;update,versions=v1alpha1,name=mnewresource.kb.io,admissionReviewVersions=v1
```

Implement `admission.Defaulter[*trainer.NewResource]`:

```go
type NewResourceDefaulter struct{ ... }

func (d *NewResourceDefaulter) Default(ctx context.Context, obj *trainer.NewResource) error { ... }
```

### Validation webhook

```go
// +kubebuilder:webhook:path=/validate-trainer-kubeflow-org-v1alpha1-newresource,mutating=false,failurePolicy=fail,sideEffects=None,groups=trainer.kubeflow.org,resources=newresources,verbs=create;update,versions=v1alpha1,name=vnewresource.kb.io,admissionReviewVersions=v1
```

Implement `admission.Validator[*trainer.NewResource]`:

```go
type NewResourceValidator struct{ ... }

func (v *NewResourceValidator) ValidateCreate(ctx context.Context, obj *trainer.NewResource) (admission.Warnings, field.ErrorList) { ... }
func (v *NewResourceValidator) ValidateUpdate(ctx context.Context, oldObj, newObj *trainer.NewResource) (admission.Warnings, field.ErrorList) { ... }

> func (v *NewResourceValidator) ValidateDelete(ctx context.Context, obj *trainer.NewResource) (admission.Warnings, field.ErrorList) { ... }

Reviewer: `admission.Validator` requires `error`, not `field.ErrorList`. Build a `field.ErrorList` in a helper and `.ToAggregate()` it at the boundary (see `trainjob_webhook.go:110`).

```suggestion
func (v *NewResourceValidator) ValidateCreate(ctx context.Context, obj *trainer.NewResource) (admission.Warnings, error) { ... }
func (v *NewResourceValidator) ValidateUpdate(ctx context.Context, oldObj, newObj *trainer.NewResource) (admission.Warnings, error) { ... }
func (v *NewResourceValidator) ValidateDelete(ctx context.Context, obj *trainer.NewResource) (admission.Warnings, error) { ... }
```

```

### Register with manager

```go

> func setupNewResourceWebhook(mgr ctrl.Manager) error {

Reviewer: Repo convention is `setupWebhookFor<Kind>`:

```suggestion
func setupWebhookForNewResource(mgr ctrl.Manager) error {
```

    return ctrl.NewWebhookManagedBy(mgr, &trainer.NewResource{}).
        WithDefaulter(&NewResourceDefaulter{...}).
        WithValidator(&NewResourceValidator{...}).
        Complete()
}
```

Add the call in `pkg/webhooks/setup.go`:

```go
func Setup(mgr ctrl.Manager, runtimes map[string]runtime.Runtime) error {
    // ... existing calls ...
    if err := setupNewResourceWebhook(mgr); err != nil {
        return err
    }
    return nil

> }

Reviewer: `Setup` returns the failing Kind alongside the error (`webhooks/setup.go:26`), consumed by `main.go:166`.

```suggestion
func Setup(mgr ctrl.Manager, runtimes map[string]runtime.Runtime) (string, error) {
    // ... existing calls ...
    if err := setupWebhookForNewResource(mgr); err != nil {
        return trainer.NewResourceKind, err
    }
    return "", nil
}
```

// +kubebuilder:rbac:groups=trainer.kubeflow.org,resources=newresources/status,verbs=get;update;patch
// +kubebuilder:rbac:groups=trainer.kubeflow.org,resources=newresources/finalizers,verbs=update
```

### Reconciler struct

```go
type NewResourceReconciler struct {
    log      logr.Logger
    client   client.Client
    recorder events.EventRecorder
}

var _ reconcile.Reconciler = (*NewResourceReconciler)(nil)
```

### SetupWithManager

```go
func (r *NewResourceReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
    return builder.TypedControllerManagedBy[reconcile.Request](mgr).
        Named("newresource_controller").
        WithOptions(options).
        WatchesRawSource(source.TypedKind(mgr.GetCache(), &trainer.NewResource{}, ...)).
        Complete(r)
}
```

### Register in setup.go

File: `pkg/controller/setup.go`

Add the controller to `SetupControllers()`:

```go
newResourceReconciler := NewNewResourceReconciler(client, recorder)
if err := newResourceReconciler.SetupWithManager(mgr, options); err != nil {
    return fmt.Errorf("setting up NewResource controller: %w", err)

> }

Reviewer: `SetupControllers` returns `(string, error)`; deps come from the manager and the error isn't wrapped (`controller/setup.go:27-50`). Note `GetEventRecorder`, not `GetEventRecorderFor`.

```suggestion
if err := NewNewResourceReconciler(
    mgr.GetClient(),
    mgr.GetEventRecorder("trainer-newresource-controller"),
).SetupWithManager(mgr, options); err != nil {
    return trainer.NewResourceKind, err
}
```

```

Reference: [pkg/controller/trainjob_controller.go](../../../pkg/controller/trainjob_controller.go) (full controller pattern)

## Step 6: Add testing wrappers

File: `pkg/util/testing/wrapper.go`

Add builder wrappers following the existing patterns (e.g.,
`MakeTrainJobWrapper`, `MakeTrainingRuntimeWrapper`). These are used by
both unit and integration tests.

## Step 7: Add integration tests

### Controller tests

File: `test/integration/controller/<newresource>_controller_test.go`

> 
> Pattern:
> - Use `ginkgo.Ordered` container
> - `BeforeAll`: call `fwk.Init()` + `fwk.RunManager(cfg, true)`
> - `BeforeEach`: create a fresh namespace
> - Use builder wrappers from `pkg/util/testing`
> - Assert expected child resource creation and status conditions

Reviewer: Missing `AfterAll` — without `fwk.Teardown()` the envtest control plane leaks between suites. Also `RunManager` returns `(context.Context, client.Client)`.

```suggestion
Pattern:
- Use `ginkgo.Ordered` container
- `BeforeAll`: `fwk = &framework.Framework{}`, `cfg = fwk.Init()`, `ctx, k8sClient = fwk.RunManager(cfg, true)`
- `AfterAll`: `fwk.Teardown()` (required — otherwise the envtest control plane leaks)
- `BeforeEach`: create a fresh namespace
- Use builder wrappers from `pkg/util/testing`
- Assert expected child resource creation and status conditions
```

### Webhook tests

File: `test/integration/webhooks/<newresource>_test.go`

Pattern:
- `BeforeAll`: call `fwk.Init()` + `fwk.RunManager(cfg, false)` (no controllers)

> - Test validation rejection and defaulting

Reviewer: Majority convention is `_webhook_test.go` (`trainjob_test.go` predates it). Also the `false` arg only gates controllers — `RunManager` always calls `webhooks.Setup` (`framework.go:120`), which is easy to misread as "webhooks off".

```suggestion
File: `test/integration/webhooks/<newresource>_webhook_test.go`

Pattern:
- `BeforeAll`: `cfg = fwk.Init()`, `ctx, k8sClient = fwk.RunManager(cfg, false)` — `false` skips controllers only; webhooks are always installed
- `AfterAll`: `fwk.Teardown()`
- Test validation rejection and defaulting
```

Reference:
- [test/integration/controller/trainjob_controller_test.go](../../../test/integration/controller/trainjob_controller_test.go)
- [test/integration/webhooks/trainjob_test.go](../../../test/integration/webhooks/trainjob_test.go)

The integration test framework (`test/integration/framework/framework.go`)
automatically loads CRDs from `manifests/base/crds/` and webhook configs
from `manifests/base/webhook/manifests.yaml`.

## Step 8: Generate manifests and verify

```bash
make manifests
```

This generates:
- CRD YAML in `manifests/base/crds/trainer.kubeflow.org_newresources.yaml`
- Updated RBAC in `manifests/base/rbac/role.yaml`
- Updated webhook config in `manifests/base/webhook/manifests.yaml`

> - CRD copied to `charts/kubeflow-trainer/templates/crd/` with Helm conditionals

Reviewer: Two manual steps are missing. Evidence they get missed: `optimizationjobs` is absent from `manifests/base/crds/kustomization.yaml` today.

```suggestion
This generates:
- CRD YAML in `manifests/base/crds/trainer.kubeflow.org_newresources.yaml`
- Updated RBAC in `manifests/base/rbac/role.yaml`
- Updated webhook config in `manifests/base/webhook/manifests.yaml`
- CRD copied to `charts/kubeflow-trainer/templates/crd/` with Helm conditionals

Two steps `make manifests` does **not** do — add them by hand:
- List the new CRD file in `manifests/base/crds/kustomization.yaml`
- Add the new resource's rules to `charts/kubeflow-trainer/templates/rbac/clusterrole.yaml` (the chart's RBAC is hand-maintained; only CRDs are copied in)
```

Then run the full validation:

```bash
make fmt
make test
make test-integration
make golangci-lint
```

## Verification

After all steps, confirm:
- `make generate` produces no uncommitted diff (generated files are up to date)
- `make manifests` produces no uncommitted diff
- `make test` passes
- `make test-integration` passes
- `make golangci-lint` passes
- New CRD YAML exists in `manifests/base/crds/` and `charts/kubeflow-trainer/templates/crd/`

> - No unrelated files were modified

Reviewer: Missing several checks CI actually runs, so a PR passing this list can still go red. `make generate` already covers `manifests`, so that line is redundant.

```suggestion
- `make generate` produces no uncommitted diff (generated files are up to date)
- `go mod tidy` produces no uncommitted diff
- `make fmt` and `make vet` produce no uncommitted diff
- `make verify-boilerplate` passes (license headers on all new files)
- `make test` passes
- `make test-integration` passes
- `make golangci-lint` passes
- `make helm-unittest` passes (chart CRDs changed)
- `make test-python` passes (`api/python_api/` models are regenerated for a new CRD)
- New CRD YAML exists in `manifests/base/crds/` and `charts/kubeflow-trainer/templates/crd/`
- No unrelated files were modified
```
